lib/cocoapods/resolver.rb

```diff
@@ -91,12 +124,12 @@ def specs_by_target
           dependencies = {}
           specs = target.dependencies.map(&:name).flat_map do |name|
             node = @activated.vertex_named(name)
-            valid_dependencies_for_target_from_node(target, dependencies, node) << node
+            (valid_dependencies_for_target_from_node(target, dependencies, node) << node).map { |s| [s, node.payload.test_specification?] }
```

Comment:
this is the "meat" of the change. We basically wrap all paylods with a `ResolvedSpecification` container and as we walk through the tree we ensure all nodes are marked as to whether this was for a test spec or not.